[MLIR][NVVM] Update TMA Load Op

durga4github opened the pull request:
This patch includes im2col and gather mode
support for the TMA Load Op. The lowering is
also updated to intrinsics except when a Predicate
is given. This completes the Blackwell additions
on this Op.

* NVVM Dialect has support for Shared::Cluster
   address-space now. So, this patch also updates the
   Op to use AS(7) instead of AS(3). The corresponding
   inline-ptx based unit tests are also updated.
*  lit tests are added for all combinations.

grypp commented on mlir/lib/Dialect/LLVMIR/IR/NVVMDialect.cpp:
```diff
@@ -157,6 +150,17 @@ LogicalResult CpAsyncBulkTensorPrefetchOp::verify() {
+  if (getPredicate()) {
+    if (mode != TMALoadMode::TILE && mode != TMALoadMode::IM2COL)
+      return emitError(
+          "Inline-ptx lowering supported only for Tile/Im2col mode.");
```
Let's change the error message `Predicate isn't supported ....`

durga4github replied:
Sure, will update.

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td:
```diff
@@ -2367,6 +2367,23 @@ def TMAStoreModeAttr : EnumAttr<NVVM_Dialect, TMAStoreMode, "tma_store_mode"> {
+  let genSpecializedAttr = 0;
+  let cppNamespace = "::mlir::NVVM";
+}
+def Tcgen05GroupKindAttr :
```
Can we rename this to `CTAGroupKindAttr`. I think TMA lives beyond tcgen05.

durga4github replied:
Addressed with PR 156448. So, resolving this.

durga4github committed:
[MLIR][NVVM][v3] Update TMA Load Op

This patch includes im2col and gather mode
support to the TMA Load Op. The lowering is
also updated to intrinsics except when Predicate
is given. This completes the Blackwell additions
on this Op.

lit tests are added for all combinations.

Signed-off-by: Durgadoss R <durgadossr@nvidia.com>

durga4github commented on mlir/lib/Dialect/LLVMIR/IR/NVVMDialect.cpp:
```diff
@@ -1535,6 +1560,123 @@ mlir::NVVM::IDArgPair CpAsyncBulkSharedCTAToGlobalOp::getIntrinsicIDAndArgs(
+  args.push_back(mt.lookupValue(thisOp.getTmaDescriptor()));
+
+  // Coordinates and im2col-offsets
+  for (auto v : thisOp.getCoordinates())
```
Sure, fixed.

Resolving this,

durga4github commented on mlir/lib/Dialect/LLVMIR/IR/NVVMDialect.cpp:
```diff
@@ -1535,6 +1560,123 @@ mlir::NVVM::IDArgPair CpAsyncBulkSharedCTAToGlobalOp::getIntrinsicIDAndArgs(
+  }
+
+  const unsigned NI = llvm::Intrinsic::not_intrinsic;
+  static constexpr llvm::Intrinsic::ID IDTable[][6] = {
```
Done.

durga4github commented on mlir/lib/Dialect/LLVMIR/IR/NVVMDialect.cpp:
```diff
@@ -1553,6 +1578,131 @@ mlir::NVVM::IDArgPair CpAsyncBulkSharedCTAToGlobalOp::getIntrinsicIDAndArgs(
+    args.push_back(builder.getInt1(hasCacheHint));
+  }
+
+  constexpr unsigned notIntrinsic = llvm::Intrinsic::not_intrinsic;
```
Resolving this,
